Add the property background repair strategy documentation (#702)
apache/skywalking-banyandb: docs/concept/property-repair.md

Unlike typical gossip-based message dissemination, where messages are spread randomly, the number of **data nodes** in BanyanDB is fixed and relatively small.
To ensure greater stability, the random peer selection is replaced with a deterministic strategy, where each node communicates only with its next node in the sequence.
Additionally, to minimize network traffic, only one peer node is involved in each round of communication.

BanyanDB already has a built-in [cluster discovery and data transmission](./clustering.md) mechanism. Therefore, the implementation of gossip protocol can be built as an extension on top of the existing cluster protocol.
Since each node already exposes a gRPC port, there is no need to introduce any additional ports, simplifying the deployment and integration.

### Propagation Message

When initiating gossip message propagation, the sender node must include both the list of participating nodes and the message content. 
The gossip protocol then proceeds through the following steps:

> 1: **Build the Context**: A context object is attached to each message and includes the following parameters:

Reviewer: Which node builds the context? The first node?

7. B node received the result, and the gossip propagation is completed.

### Tracing

Tracing becomes critically important in a gossip-based protocol, as issues can arise at any stage of the communication and synchronization process.
New trace spans are created in the following scenarios:

1. **Initiating Node**: The initiating node records the full trace of the entire request, capturing all steps from message dispatch to the final result collection.
2. **Receiving Sync Requests**: When a node receives a sync request and begins communication with another peer node, it creates a new trace span for that interaction to track the request lifecycle independently.
3. **Business Logic Execution**: During the actual business processing (e.g., data comparison, update, and transmission), custom trace data and metadata can be attached to the active trace context to provide deeper insights into logic-specific behaviors.

After each synchronization cycle, the receiving node will send its trace data back to the initiating node, allowing the initiator to aggregate and correlate all spans for end-to-end visibility and debugging.

## Property Repair

> Once all the above preparations are complete, the system can proceed with the Property Repair process.

Reviewer: What's the "preparations"? The previous section is "Gossip Protocol," which is not an action.

> This process is scheduled to run on each data node daily at 1:00 AM, and follows these steps:

Reviewer: The "process" is triggered on which node?

# Property Background Repair Strategy

In Property, the `replicas` configuration allows Property data to be stored across multiple data nodes, providing high availability. 
However, during data updates, some nodes may become temporarily unavailable, which can lead to version inconsistencies of the same data across different machines.

> This documentation introduces it how the system automatically synchronizes Property data across all data nodes in the background, 

Reviewer: ```suggestion
This documentation explains how the system automatically synchronizes Property data across all data nodes in the background, 
```

2. **Snapshot the Shard**: A snapshot of the shard data is taken to avoid blocking ongoing business operations during data traversal.
3. **Build the Tree**: Using the streaming method, the system scans all data in the snapshot and builds a Merkle Tree for each group individually.
4. **Save the Snapshot ID**: The snapshot ID used in this build is saved, so it can be used for efficient change detection during the next scheduled run.

## Gossip Protocol

In BanyanDB's typical data communication model, data exchange is primarily handled by **liaison** nodes, which interact with **data nodes** through **broadcasting or publishing** mechanisms.
However, in the context of Property data repair, involving liaison nodes would introduce unnecessary network overhead. 
Therefore, the system adopts the gossip protocol, allowing data nodes to communicate directly with each other to perform data repair in a more efficient and decentralized manner.

Unlike typical gossip-based message dissemination, where messages are spread randomly, the number of **data nodes** in BanyanDB is fixed and relatively small.

> To ensure greater stability, the random peer selection is replaced with a deterministic strategy, where each node communicates only with its next node in the sequence, and the node list is ordered. 

Reviewer: Node B has crashed or is unreachable. What happens now? The document doesn't specify. Does Node A wait forever? Does it time out? Does the entire repair for the group fail? A robust protocol should be able to skip the dead node and continue the process with the next available one (e.g., A syncs with C). As described, the crash of a single node can halt the entire repair chain for that round.

3. **Wait for the Result**: The initiating node waits for the final result of the synchronization process before proceeding.

### Property Synchronize between Two Nodes

When two nodes engage in Property data synchronization, they follow a specific protocol to ensure data consistency.
Let’s refer to the current node as A and the target node as B. The process is as follows:

1. **(A)Establish Connection**: Node A initiates a **bidirectional streaming connection** with node B to enable real-time, two-way data transfer.
2. **(A)Iterate Over Shards**: Node A retrieves the list of all Property-related shards for the selected group and processes them one by one.
3. **(A)Send Merkle Tree Summary**: For each shard, node A reads its Merkle Tree and sends a summary (including root SHA and slots SHA) to node B. 
This allows B to quickly identify which slots may contain differences.
4. **(B)Verify Merkle Tree Summary and Respond**: Node B compares the received summary against its own Merkle Tree for the same shard and group:

>    * If the root SHA match, node B returns an empty slot list, indicating no differences. 

Reviewer: ```suggestion
   * If the root SHA matches, node B returns an empty slot list, indicating no differences. 
```

   * If the root SHA differ, node B checks the slot SHA, identifies mismatched slots, and sends back all relevant leaf node details, including the **slot index**, **entity**, and **SHA value**.
5. **(A)Compare Leaf Data**: Node A processes the received leaf data and takes the following actions: 
   * For missing entities (present on B but not on A), A requests the full Property data from B.
   * For entities present on A, but not on B, A sends the full Property data to B.
   * For SHA mismatches, A sends its full Property data to B for validation.
6. **(B)Validate Actual Data**: Node B handles the data as follows: 
   * For missing entities, B returns the latest version of the data.
   * For entities present on A, but not on B, B updates its local copy with the data from A.

>    * For SHA mismatches, B compares the version numbers, If B’s version is newer, it returns the Property data to A. If A’s version is newer, B updates its local copy and does not return any data.

Reviewer: ```suggestion
   * For SHA mismatches, B uses "last-write-win" strategy. It compares the version numbers. If B’s version is newer, it returns the Property data to A. If A’s version is newer, B updates its local copy and does not return any data. If the versions are the same, it selects the data from the smaller index of the node list; in this case, it would be from A.
```

7. **(A)Update Local Data**: Node A receives updated Property data from B and applies the changes to its local store.

This concludes the A-to-B synchronization cycle for a given group in the Property repair process.

### Error Handling

During the synchronization process, the system will terminate or skip processing under the following scenarios:

1. **Merkle Tree Not Built**: If either node A or node B has not yet built the Merkle Tree for the target group, the gossip protocol is immediately terminated.
2. **Duplicate Sync Request**: If a new gossip sync request for the same group is received by either node while an existing synchronization is in progress, the new request is terminated to avoid conflicts.
3. **Target Node Request Failure**: If node B fails to send or respond to requests during the sync process, the gossip protocol is terminated.

> 4. **Property Repair Failure**: If an error occurs while applying Property data updates (e.g., write or query failure). Then, record the error info into the trace and ignore current property sync between these two nodes and keep sync continue.

Reviewer: ```suggestion
4. **Property Repair Failure**: If an error occurs while applying Property data updates (e.g., a write failure), the error is recorded in the trace. The synchronization for that specific property is skipped, but the overall gossip process continues with the next item
```

5. **Unhandled Exceptions**: For any other unexpected exceptions or failures, the gossip protocol is immediately terminated to maintain system consistency and avoid cascading errors.

## Background Repair vs. Query Repair

> When querying Property data, the liason node will still query the latest available data from existing data nodes, 

Reviewer: ```suggestion
When querying Property data, the liaison node will still query the latest available data from existing data nodes, 
```

This allows B to quickly identify which slots may contain differences.
4. **(B)Verify Merkle Tree Summary and Respond**: Node B compares the received summary against its own Merkle Tree for the same shard and group:
   * If the root SHA matches, node B returns an empty slot list, indicating no differences. 
   * If the root SHA differ, node B checks the slot SHA, identifies mismatched slots, and sends back all relevant leaf node details, including the **slot index**, **entity**, and **SHA value**.
5. **(A)Compare Leaf Data**: Node A processes the received leaf data and takes the following actions: 
   * For missing entities (present on B but not on A), A requests the full Property data from B.
   * For entities present on A, but not on B, A sends the full Property data to B.
   * For SHA mismatches, A sends its full Property data to B for validation.
6. **(B)Validate Actual Data**: Node B handles the data as follows: 
   * For missing entities, B returns the latest version of the data.
   * For entities present on A, but not on B, B updates its local copy with the data from A.

>    * For SHA mismatches, B uses "last-write-win" strategy. It compares the version numbers. If B’s version is newer, it returns the Property data to A. If A’s version is newer, B updates its local copy and does not return any data. If the versions are the same, it selects the data from the smaller index of the node list; in this case, it would be from A.****

Reviewer: ```suggestion
   * For SHA mismatches, B uses the "last-write-win" strategy. It compares the version numbers. If B’s version is newer, it returns the Property data to A. If A’s version is newer, B updates its local copy and does not return any data. If the versions are the same, it selects the data from the smaller index of the node list; in this case, it would be from A.
```
